atomics/include/desul/atomics/Fetch_Op_HIP.hpp

```diff
@@ -9,9 +9,90 @@ SPDX-License-Identifier: (BSD-3-Clause)
 #ifndef DESUL_ATOMICS_FECH_OP_HIP_HPP_
 #define DESUL_ATOMICS_FECH_OP_HIP_HPP_
 
+#include "desul/atomics/Adapt_GCC.hpp"
+
+#if __has_builtin(__hip_atomic_fetch_add) && __has_builtin(__hip_atomic_fetch_min) && \
+    __has_builtin(__hip_atomic_fetch_max)
+// The above intrinsics are the minimum set needed to implement the math
+// operators:
+//  fetch_add, fetch_sub, fetch_min, fetch_max, fetch_inc, fetch_dec
+#define DESUL_HAVE_HIP_ATOMIC_MATH_DETAIL
+#endif
+#if __has_builtin(__hip_atomic_fetch_and) && __has_builtin(__hip_atomic_fetch_or) && \
+    __has_builtin(__hip_atomic_fetch_xor)
+// The above intrinsics are needed for bitwise operations on integral types.
+#define DESUL_HAVE_HIP_ATOMIC_BIT_DETAIL
+#endif
+
+#if defined(DESUL_HAVE_HIP_ATOMIC_BIT_DETAIL) || \
+    defined(DESUL_HAVE_HIP_ATOMIC_MATH_DETAIL)
+#include "desul/atomics/Adapt_HIP.hpp"
+#endif
+
 namespace desul {
 namespace Impl {
 
+#ifdef DESUL_HAVE_HIP_ATOMIC_MATH_DETAIL
+template <typename T, typename MemoryOrder, typename MemoryScope>
+inline __device__ std::enable_if_t<std::is_arithmetic<T>::value, T>
+device_atomic_fetch_add(T* ptr, T val, MemoryOrder, MemoryScope) {
+  return __hip_atomic_fetch_add(
+      ptr, val, GCCMemoryOrder<MemoryOrder>::value, HIPMemoryScope<MemoryScope>::value);
+}
+
+template <typename T, typename MemoryOrder, typename MemoryScope>
+inline __device__ std::enable_if_t<std::is_arithmetic<T>::value, T>
+device_atomic_fetch_sub(T* ptr, T val, MemoryOrder, MemoryScope) {
+#if __has_builtin(__hip_atomic_fetch_sub)
+  return __hip_atomic_fetch_sub(
+      ptr, val, GCCMemoryOrder<MemoryOrder>::value, HIPMemoryScope<MemoryScope>::value);
+#else
+  return __hip_atomic_fetch_add(ptr,
+                                -val,
+                                GCCMemoryOrder<MemoryOrder>::value,
+                                HIPMemoryScope<MemoryScope>::value);
+#endif
+}
+
+template <typename T, typename MemoryOrder, typename MemoryScope>
+inline __device__ std::enable_if_t<std::is_arithmetic<T>::value, T>
+device_atomic_fetch_min(T* ptr, T val, MemoryOrder, MemoryScope) {
+  return __hip_atomic_fetch_min(
+      ptr, val, GCCMemoryOrder<MemoryOrder>::value, HIPMemoryScope<MemoryScope>::value);
+}
+
+template <typename T, typename MemoryOrder, typename MemoryScope>
+inline __device__ std::enable_if_t<std::is_arithmetic<T>::value, T>
+device_atomic_fetch_max(T* ptr, T val, MemoryOrder, MemoryScope) {
+  return __hip_atomic_fetch_max(
+      ptr, val, GCCMemoryOrder<MemoryOrder>::value, HIPMemoryScope<MemoryScope>::value);
+}
```

Comment:
Did you test these guys?
I see that I had commented the fp overloads on my version https://github.com/kokkos/kokkos/pull/6031/files#diff-aae30b453c4ca28bccc8e0a5c46d3c1250b78cce5b3187ecb869cca34339e169R44-R45 which I suppose mean I had seen issues.  Of course I haven't taken notes and I don't remember what the deal was.

Reply:
I ran through the `test-forall-atomic-basic-Hip.exe` in RAJA, built with `RAJA_ENABLE_EXHAUSTIVE=ON`.

Looking at the chunks of testing code [here](https://github.com/LLNL/RAJA/blob/a669daa06253d076e8ed57d35c55fd7a9e8c8614/test/include/RAJA_test-atomic-types.hpp#L24-L33) and [here](https://github.com/LLNL/RAJA/blob/a669daa06253d076e8ed57d35c55fd7a9e8c8614/test/functional/forall/atomic-basic/tests/test-forall-atomic-basic.hpp#L103-L112), running that test should exercise `atomicMin` and `atomicMax` for all types, including `float` and `double`.

Reply:
Please confirm what version(s) of ROCm you tested.

Reply:
I tested 5.7.1 and 6.0.2. Are there other versions you'd want to test?

Reply:
On the Kokkos end we need to support back to 5.2